lib/Analysis/Allocation.cpp

```diff
@@ -595,45 +714,73 @@ class AllocationAnalysis {
                 DenseMap<BufferT *, size_t> &bufferStart) {
     // Reset shared memory size
     allocation->sharedMemorySize = 0;
+
+    SmallVector<BufferT *> xBuffers;
+    for (auto *buf : buffers) {
+      if (buf->kind != BufferT::BufferKind::Alias)
+        xBuffers.push_back(buf);
+    }
+
     // First-fit graph coloring
     // Neighbors are nodes that interfere with each other.
     // We color a node by finding the index of the first available
     // non-neighboring node or the first neighboring node without any color.
     // Nodes with the same color do not interfere with each other.
     DenseMap<BufferT *, int> colors;
-    for (auto value : buffers) {
-      colors[value] = (value == buffers[0]) ? 0 : -1;
+    for (auto value : xBuffers) {
+      colors[value] = (value == xBuffers.front()) ? 0 : -1;
     }
-    SmallVector<bool> available(buffers.size());
-    for (auto x : buffers) {
+    SmallVector<bool> available(xBuffers.size());
+    for (auto x : xBuffers) {
       std::fill(available.begin(), available.end(), true);
       for (auto y : interference.lookup(x)) {
-        int color = colors[y];
-        if (color >= 0) {
-          available[color] = false;
-        }
+        y->apply([&](BufferT *buf) {
+          if (buf != x) {
+            int color = colors[buf];
+            if (color >= 0)
+              available[color] = false;
+          }
+        });
       }
-      auto it = std::find(available.begin(), available.end(), true);
+      auto it = llvm::find(available, true);
       colors[x] = std::distance(available.begin(), it);
+      LLVM_DEBUG(llvm::dbgs()
+                 << "    COLOR: " << x->id << " -> " << colors[x] << "\n");
     }
     // Finalize allocation
     // color0: [0, 7), [0, 8), [0, 15) -> [0, 7), [0, 8), [0, 15)
     // color1: [7, 9) -> [0 + 1 * 15, 9 + 1 * 15) -> [15, 24)
     // color2: [8, 12) -> [8 + 2 * 15, 12 + 2 * 15) -> [38, 42)
     // TODO(Keren): We are wasting memory here.
     // Nodes with color2 can actually start with 24.
-    for (auto x : buffers) {
+    for (auto x : xBuffers) {
       size_t adj = 0;
       for (auto y : interference.lookup(x)) {
-        adj = std::max(adj, bufferStart.lookup(y) + y->size);
+        y->apply([&](BufferT *buf) {
+          if (buf != x)
+            adj = std::max(adj, buf->size);
+        });
       }
+      // TODO(SJW): does not take alignment into account
```

Comment:
The new offset of the Buffer could be `adjusted` by the `size` of the other Buffer which could violate `alignment`.

Reply:
hmm, note that the allocation analysis is supposed to succeed in any case.

I get the point:

```
    const size_t scratchAlignment = 128;
    size_t kAlignment = 8;
```

Because these two alignments are different, it's possible that alignment is violated.

Maybe we could revert it back to `bufferStart.lookup(y) + y->size` for now?

Reply:
I think we may break the alignment still. Maybe a better way is to realign the memory address

Reply:
`adj = std::max(adj, buf->size);` this could be kept.

Then we realign the address by modifying `x->offset` based on its alignment. What do you think?

Reply:
Since `x` starts aligned, we could just bump by `x->alignment` every time until there is no conflict.

Reply:
hmm, I'm not sure if that'll take much longer time. Let's be a bit more aggressive for now

Reply:
yes, this did result in several more iterations.

I think instead of incrementing the current `x->offset` with the bound of all interference buffers, we can just take the bound, on the next nearest alignment.